controller/src/main/java/io/pravega/controller/store/stream/BucketStore.java

```diff
@@ -0,0 +1,129 @@
+     *
+     * @param serviceType service type
+     * @param listener    listener
+     */
+    void registerBucketOwnershipListener(ServiceType serviceType, BucketOwnershipListener listener);
+
+    /**
+     * Unregister listeners for bucket ownership.
+     *
+     * @param serviceType service type
+     */
+    void unregisterBucketOwnershipListener(ServiceType serviceType);
+
+    /**
+     * Method to take ownership of a bucket.
+     *
+     * @param serviceType service type
+     * @param bucket      bucket id
+     * @param processId   process id
+     * @param executor    executor
+     * @return future, which when completed, will contain a boolean which tells if ownership attempt succeeded or failed.
+     */
+    CompletableFuture<Boolean> takeBucketOwnership(ServiceType serviceType, int bucket, String processId, Executor executor);
+
+    // region retention
+
+    /**
+     * Method to register listeners for changes to streams under the bucket.
+     *
+     * @param serviceType service type
+     * @param bucket      bucket
+     * @param listener    listener
+     */
+    void registerBucketChangeListener(ServiceType serviceType, int bucket, BucketChangeListener listener);
+
+    /**
+     * Method to unregister listeners for changes to streams under the bucket.
+     *
+     * @param serviceType service type
+     * @param bucket      bucket
+     */
+    void unregisterBucketChangeListener(ServiceType serviceType, int bucket);
+
+    /**
+     * Return all streams in the bucket.
+     *
+     * @param serviceType service type
+     * @param bucket      bucket id.
+     * @param executor    executor
+     * @return Future, which when completed will have a list of scopedStreamName (scope/stream) of all streams under the bucket.
+     */
+    CompletableFuture<List<String>> getStreamsForBucket(ServiceType serviceType, int bucket, Executor executor);
+
+    /**
+     * Add the given stream to appropriate bucket for auto-retention.
+     *
+     * @param serviceType service type
+     * @param scope       scope
+     * @param stream      stream
+     * @param executor    executor
+     * @return future, which when completed will indicate that stream is added to bucket store.
+     */
+    CompletableFuture<Void> addStreamToBucketStore(ServiceType serviceType, String scope, String stream, Executor executor);
+
+    /**
+     * Remove stream from auto retention bucket.
+     *
+     * @param serviceType service type
+     * @param scope       scope
+     * @param stream      stream
+     * @param executor    executor
+     * @return future, which when completed will indicate that stream is removed from the store.
+     */
+    CompletableFuture<Void> removeStreamFromBucketStore(ServiceType serviceType, String scope, String stream, Executor executor);
+    // endregion
+
+    enum ServiceType {
+        // Naming the service id as "buckets" for backward compatibility
+        RetentionService("buckets"),;
```

Comment:
Has it been considered a scenario with Controller having different versions of this `ServiceType` enum?

Reply:
Service type is an internal class. 

Versioning is relevant on Bucket store's implementation for any serialized data. 
In ZkBucketStore, we presently only follow a path hierarchy and not make use of any serialized blob for versioning to come into play there.